src/core/or/circuitpadding.c

```diff
@@ -0,0 +1,2370 @@
+          && mi->histogram_len != state->histogram_len)) {
+    tor_free(mi->histogram); // null ok
+    mi->histogram = tor_malloc_zero(sizeof(circpad_hist_bin_t)
+                                    *state->histogram_len);
+  }
+  mi->histogram_len = state->histogram_len;
+
+  memcpy(mi->histogram, state->histogram,
+         sizeof(circpad_hist_bin_t)*state->histogram_len);
+}
+
+/**
+ * Choose a length for this state (in cells), if specified.
+ */
+static void
+circpad_choose_state_length(circpad_machineinfo_t *mi)
+{
+  const circpad_state_t *state = circpad_machine_current_state(mi);
+  double length;
+
+  if (!state || state->length_dist.type == CIRCPAD_DIST_NONE) {
+    mi->state_length = CIRCPAD_STATE_LENGTH_INFINITE;
+    return;
+  }
+
+  length = circpad_distribution_sample(state->length_dist);
+  length = MAX(0, length);
+  length += state->start_length;
+  length = MIN(length, state->max_length);
+
+  mi->state_length = clamp_double_to_int64(length);
+}
+
+/**
+ * Sample an expected time-until-next-packet delay from the histogram.
+ *
+ * The bin is chosen with probability proportional to the number
+ * of tokens in each bin, and then a time value is chosen uniformly from
+ * that bin's [start,end) time range.
+ */
+static circpad_delay_t
+circpad_machine_sample_delay(circpad_machineinfo_t *mi)
+{
+  const circpad_state_t *state = circpad_machine_current_state(mi);
+  const circpad_hist_bin_t *histogram = NULL;
+  int i = 0;
+  uint32_t curr_weight = 0;
+  uint32_t histogram_total = 0;
+  uint32_t bin_choice;
+  circpad_delay_t bin_start, bin_end;
+  circpad_delay_t start_usec;
+
+  tor_assert(state);
+
+  if (state->use_rtt_estimate)
+    start_usec = mi->rtt_estimate_us+state->start_usec;
+  else
+    start_usec = state->start_usec;
+
+  if (state->token_removal != CIRCPAD_TOKEN_REMOVAL_NONE) {
+    /* We have a mutable histogram */
+    tor_assert(mi->histogram && mi->histogram_len == state->histogram_len);
+
+    histogram = mi->histogram;
+    for (int b = 0; b < state->histogram_len; b++)
+      histogram_total += histogram[b];
+  } else if (state->iat_dist.type == CIRCPAD_DIST_NONE) {
+    /* We have a histogram, but it's immutable */
+    histogram = state->histogram;
+    histogram_total = state->histogram_total;
+  } else {
+    /* Sample from a fixed IAT distribution and return */
+    double val = circpad_distribution_sample(state->iat_dist);
+    /* These comparisons are safe, because the output is in the range
+     * [0, 2**46], and double has a precision of 53 bits. */
+    val = MAX(0, val);
+    val = MIN(val, state->range_sec*TOR_USEC_PER_SEC);
+
+    /* This addition is exact: val is at most 2**16 * USEC_PER_SEC ~= 2**46,
+     * start_usec is at most 2**32, and doubles have a precision of 53 bits. */
+    val += start_usec;
+    /* Clamp the distribution at infinite delay val */
+    return (circpad_delay_t)MIN(tor_llround(val), CIRCPAD_DELAY_INFINITE);
+  }
+
+  bin_choice = crypto_rand_int(histogram_total);
```

Comment:
Suggestion: Let's rename `histogram_total` to `histogram_total_tokens` to make it more explicit.

In this function it's not clear whether `histogram_total` is the number of bins or the number of tokens, and actually the above line strongly implies the former when it's the latter.

Reply:
All the variable changing you guys asked for in this and other comments is done across these three commits (so as to squash sanely): d8c4eff8c5d62d172eac0d5096f32f0be99ce6cf 979cc31a3e3e97619760c7aa4f94c8f767cdd29a e18d5a3111375b995a57e809f080011d6ebe5ad5